### PR TITLE
generator: Convert to pathlib, fix wrong path passed to various executables

### PR DESCRIPTION
**Pull request by MarijnS95**
`def_check_submodule` was always passing `'gir'` even when called with `gir-files`, and the `depth` used to derive whether `Gir.toml` resides in a `sys` crate is inaccurate when specifying eg. `**/sys` on the command line (as that'd be level 0).

In addition, convert path handling to `pathlib` whose paths are constructed and checked during argument parsing.

Please double-check if I overlooked a path anywhere that's still a string instead of `Path`.

EDIT: With pathlib we might also replace the depth-first search in `regen_crates` with `path.glob` or `path.rglob` on `**/Gir*.toml`?

### CODE REVIEW

**Comment by sdroege** [comment]
CC @sophie-h

**Comment by MarijnS95** [comment]
Yeah, thanks @sophie-h for doing the argparse conversion!  I thought about that ever since the previous revision landed on `gstreamer-rs` past week.

File: generator.py [hunk 1]
```diff
@@ -77,18 +73,16 @@ def build_gir_if_needed(updated_submodule):
...
-        if isdir(entry_file):
-            if level < 2 and not regen_crates(entry_file, conf, level + 1):
+    for entry in path.iterdir():
+        if entry.is_dir():
+            if level < 2 and not regen_crates(entry, conf, level + 1):
                 return False
-        elif entry.startswith("Gir") and entry.endswith(".toml"):
-            print('==> Regenerating "{}"...'.format(entry_file))
+        elif entry.name.startswith("Gir") and entry.suffix == ".toml":
+            print('==> Regenerating "{}"...'.format(entry))
```
Comment (GuillaumeGomez) [inline generator.py:81]: Just to confirm: it prints the "full path" right?

Comment (MarijnS95) [reply generator.py:81]: Yup, `entry.name` is used intentionally in the line above to get just the file name.

File: generator.py [hunk 2]
```diff
@@ -11,32 +10,29 @@
 NEED_UPDATE = 1
 FAILURE = 2
 
-DEFAULT_GIR_DIRECTORY = 'gir-files'
-DEFAULT_GIR_PATH = './gir/target/release/gir'
+DEFAULT_GIR_FILES = Path('./gir-files')
```
Comment (GuillaumeGomez) [inline generator.py:13]: I suggest `DEFAULT_GIR_FILES_DIRECTORY` instead to be consistent with `DEFAULT_GIR_DIRECTORY`.

File: generator.py [hunk 3]
```diff
@@ -11,32 +10,29 @@
...
-    except:
-        return False
+    return run_command(['cargo', 'build', '--release'], 'gir')
 
 
 def ask_yes_no_question(question, conf):
     question = '{} [y/N] '.format(question)
     if conf.yes:
-        print(question)
+        print(question + 'y')
```
Comment (GuillaumeGomez) [inline generator.py:35]: Why adding the `y` there?

File: generator.py (hunk 2 above)
Comment (MarijnS95) [reply generator.py:13]: Makes sense, will update :)

File: generator.py (hunk 3 above)
Comment (MarijnS95) [reply generator.py:35]: I'd rather not print this at all. It's a bit weird to see `[y/N]` (fullcaps `N` meaning default no show up) but the program continuing, with seemingly no answer given.

File: generator.py (hunk 2 above)
Comment (MarijnS95) [reply generator.py:13]: How about the command-line option, is `--gir-files` okay with you?

File: generator.py (hunk 3 above)
Comment (GuillaumeGomez) [reply generator.py:35]: Well, I guess it makes sense in a way. Let's go for that then!

File: generator.py (hunk 1 above)
Comment (GuillaumeGomez) [reply generator.py:81]: :+1:

File: generator.py (hunk 2 above)
Comment (GuillaumeGomez) [reply generator.py]: The option yes, but the field not really. Could you change it to `gir_files_path` please (same as `gir_path` :) )?

**Comment by bilelmoussaoui** [comment]
Would be nice to send a PR for the same changes to gtk4-rs if possible. Otherwise I can backport them once merged here

File: generator.py (hunk 2 above)
Comment (MarijnS95) [reply generator.py]: That name is derived from the flag, butt can be overridden with `dest=`.

On that note I don't find `--gir-path` very expressive, should we perhaps rename that to just `--gir` or `--gir-exe(cutable)`? We can also add some short-hands like `-g`, `-f`, '-y` etc.

**Comment by MarijnS95** [comment]
> Would be nice to send a PR for the same changes to gtk4-rs if possible. Otherwise I can backport them once merged here

@bilelmoussaoui Sure thing, I can apply the same commits there as soon as this passes review :)

File: generator.py (hunk 2 above)
Comment (GuillaumeGomez) [reply generator.py]: Please don't add short-hands. I voluntarily didn't add them. As for the name, I find `--gir-path` good enough but others opinions can't hurt. cc @bilelmoussaoui @sdroege

Comment (bilelmoussaoui) [reply generator.py]: --gir-path is good for me as well, but no hard opinions
@sophie-h any opinions?

Comment (MarijnS95) [reply generator.py]: Renamed the field to `gir_files_path`. I'm not using `--gir-path` yet so I have no stake in the choice :)

Comment (sdroege) [reply generator.py]: Let's add short versions of the parameters in a separate PR as this apparently is contentious and needs to be discussed, while everything else in this PR is easy. (FWIW I don't care either way)

**Review by sdroege — APPROVED** [review]

**Comment by sdroege** [comment]
Looks good to me but why is `regen_check` failing on the CI? Looks like the gir-files submodule is out of sync with the code. Does one of the changes here accidentally update the submodule automatically from `generator.py`? :)

File: generator.py (hunk 2 above)
Comment (MarijnS95) [reply generator.py]: @sdroege Fyi this is not about shorthands as much as it is about the naming of `--gir-path`. But whatever, this is a small development utility, not a public facing API :)

Comment (sdroege) [reply generator.py]: I also don't care either way for that one :)

**Comment by MarijnS95** [comment]
@sdroege The same happens in the `argparse` PR: https://github.com/gtk-rs/gtk-rs/runs/1438378458

`actions/checkout` in CI.yml is not set to `submodules: recursive` (or `true`) meaning that `./generator.py` gets to initialize them. Because `--yes` is passed, it'll always _pull_ `master` which is not what `gir-files` points to currently. Consequently, `git` wants you to be more specific about the `pull` strategy as well:

```
HEAD is now at 062b3d6 Merge pull request #995 from sdroege/remove-setvalueoptional-trait-bound
warning: Pulling without specifying how to reconcile divergent branches is
discouraged. You can squelch this message by running one of the following
commands sometime before your next pull:

  git config pull.rebase false  # merge (the default strategy)
  git config pull.rebase true   # rebase
  git config pull.ff only       # fast-forward only

You can replace "git config" with "git config --global" to set a default
preference for all repositories. You can also pass --rebase, --no-rebase,
or --ff-only on the command line to override the configured default per
invocation.
```

So:
- Does it even make sense to pass `--yes` in CI?
- Does it even make sense to have `pull -f origin master` in the generator?
  - Either this should be a manual action;
  - Or `reset`, or `--ff-only`.
- Have a separate flag for saying yes to continue after an error, and to update submodules to master.

**Comment by sdroege** [comment]
I don't think the submodules should be updated ever in the CI. That defeats the whole point.

**Comment by MarijnS95** [comment]
@sdroege Looks like it passes now. We have to sync the submodules directly from `CI.yml` otherwise the script will do the initial sync and ask for submodules, which fails: https://github.com/gtk-rs/gtk-rs/runs/1441060184?check_suite_focus=true

If we want to test this part of the codepath we might as well: `yes n | python3 generator.py`, what do you prefer?

**Comment by sdroege** [comment]
Something for @bilelmoussaoui and @GuillaumeGomez to decide :)

**Comment by GuillaumeGomez** [comment]
We could simply add a `--no` which would answer `no` to all questions. What do you think? Like that, the CI can't be stuck and we don't update anything that shouldn't be.

**Comment by MarijnS95** [comment]
@GuillaumeGomez In all honesty I feel like removing `input` altogether, in favor of explicit options: `--continue-on-error` and `--update-submodules-to-master` or something like that.

**Comment by GuillaumeGomez** [comment]
I prefer keeping the input. Generally it means that an issue occurred and I'd like to be able to keep a given state and choosing whether or not to continue depending on the error.

**Comment by bilelmoussaoui** [comment]
> @sdroege The same happens in the `argparse` PR: https://github.com/gtk-rs/gtk-rs/runs/1438378458
> 
> `actions/checkout` in CI.yml is not set to `submodules: recursive` (or `true`) meaning that `./generator.py` gets to initialize them. Because `--yes` is passed, it'll always _pull_ `master` which is not what `gir-files` points to currently, resulting in the diff. Consequently, `git` wants you to be more specific about the `pull` strategy as well:
> 
> ```
> HEAD is now at 062b3d6 Merge pull request #995 from sdroege/remove-setvalueoptional-trait-bound
> warning: Pulling without specifying how to reconcile divergent branches is
> discouraged. You can squelch this message by running one of the following
> commands sometime before your next pull:
> 
>   git config pull.rebase false  # merge (the default strategy)
>   git config pull.rebase true   # rebase
>   git config pull.ff only       # fast-forward only
> 
> You can replace "git config" with "git config --global" to set a default
> preference for all repositories. You can also pass --rebase, --no-rebase,
> or --ff-only on the command line to override the configured default per
> invocation.
> ```
> 
> So:
> 
>     * Does it even make sense to pass `--yes` in CI?
> 
>     * Does it even make sense to have `pull -f origin master` in the generator?
>       
>       * Either this should be a manual action;
>       * Or `reset`, or `--ff-only`.
> 
>     * Have a separate flag for saying yes to continue after an error, and to update submodules to master.

We use an action to checkout and clone a repository, you can pass a parameter to it and ask it to clone the repository recursively

```yaml

  generator:
    name: regen check
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          submodules: recursive
      - run: python3 generator.py --yes
      - run: git diff -R --exit-code
``` 
See https://github.com/actions/checkout for the various options you can pass to it

**Comment by MarijnS95** [comment]
> I prefer keeping the input. Generally it means that an issue occurred and I'd like to be able to keep a given state and choosing whether or not to continue depending on the error.

@GuillaumeGomez We can do both, the input for the error and the flag to forcibly update the submodule to `master`?

@bilelmoussaoui See the last commit and my suggestion to add `submodules: recursive` in _exactly_ the comment you quoted :stuck_out_tongue_winking_eye:

**Comment by sdroege** [comment]
So all good here now or ...? :)

**Review by bilelmoussaoui — APPROVED** [review]

**Comment by sdroege** [comment]
@GuillaumeGomez ?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by sdroege** [comment]
Ok, then this only needs someone to fix the merge conflicts :)

**Comment by MarijnS95** [comment]
@sdroege Someone, not me. @GuillaumeGomez [requested](https://github.com/gtk-rs/gtk-rs/pull/170#discussion_r528364147) 2 days ago to rename `DEFAULT_GIR_FILES_DIRECTORY` which I promptly did, but went ahead and made + self-approve [the same change](https://github.com/gtk-rs/gtk-rs/pull/174) a couple hours ago plus some other renames, conflicting with most commits in this PR. Not nice. Could have asked or waited until this PR landed, would have been easier for both of us.

**Comment by sdroege** [comment]
That's annoying and not very nice indeed. @GuillaumeGomez can you merge this PR then? You broke it :)

**Comment by GuillaumeGomez** [comment]
Arf sorry, completely forgot about this one. I was doing a gir change and noticed that. I'm very sorry.

If you want I can fix the conflict? Again, I'm very sorry I forgot about it...

File: generator.py [hunk 4]
```diff
@@ -147,6 +147,9 @@ def main():
     if conf.gir_files_path == DEFAULT_GIR_FILES_DIRECTORY:
         if def_check_submodule(conf.gir_files_path, conf) == FAILURE:
             return 1
+    elif not isdir(conf.gir_files_path):
+        print("`{}` dir doesn't exist. Aborting...".format(conf.gir_files_path))
+        return 1
```
Comment (MarijnS95) [inline generator.py]: Careful, this is dealt with by `def directory_path(path):` above, and should otherwise be `conf.gir_files_path.is_dir()`.

**Comment by MarijnS95** [comment]
@GuillaumeGomez No worries, if it's a genuine oversight then that's okay.  And merging is easier than rebasing, too :D

File: generator.py (hunk 4 above)
Comment (GuillaumeGomez) [reply generator.py]: Arf! Removing useless code I added back.

File: generator.py [hunk 5]
```diff
@@ -127,7 +127,7 @@ def parse_args():
     parser.add_argument('path', nargs="*", default=[Path('.')],
                         type=valid_path,
                         help='Paths in which to look for Gir.toml files')
-    parser.add_argument('--gir-files', dest="gir_files_path", default=DEFAULT_GIR_FILES_DIRECTORY,
+    parser.add_argument('--gir-files-directory', dest="gir_files_path", default=DEFAULT_GIR_FILES_DIRECTORY,
```
Comment (MarijnS95) [inline generator.py:130]: Do we want to persist using `dest="gir_files_path"` or switch to `gir_files_directory` from your change?

Comment (GuillaumeGomez) [reply generator.py:130]: I kept your wording. But both are fine for me.

**Comment by GuillaumeGomez** [comment]
And I removed the unnecessary condition I added back. I'm really sorry about this mess. As soon as CI is green, I merge. Thanks a lot for this! I discovered a lot of new things about python in this PR too. :)

**Comment by MarijnS95** [comment]
@GuillaumeGomez Cool, thanks for fixing it up! Conversely I'm learning a lot about gir/gstreamer/gtk in Rust and the support code around it, thanks for the ride yall!

**Comment by sdroege** [comment]
@MarijnS95 Thanks for the improvements :)

File: generator.py [hunk 6]
```diff
@@ -11,32 +10,29 @@
...
 
 
 def ask_yes_no_question(question, conf):
     question = '{} [y/N] '.format(question)
     if conf.yes:
-        print(question)
+        print(question + 'y')
         return True
     if sys.version_info[0] < 3:
         line = raw_input(question)
```
Comment (sdroege) [inline generator.py:38]: Oops, there is no function `raw_input()` anywhere. Glad we're using Python so nobody noticed this ;)

Comment (MarijnS95) [reply generator.py:38]: @sdroege Should be in Python 2, which is what this code falls back to?

Then again Python 2 is deprecated for almost a year now, and the shebang specifies Python 3, might as well remove this in a followup PR :)

Comment (sdroege) [reply generator.py:38]: Oh I just had my editor complain that this function doesn't exist (and I don't have Python 2 installed)

Comment (bilelmoussaoui) [reply generator.py:38]: The code uses `Path` & some new fancy python 3 features, it should be safe to remove the python 2 fallback I would say.

Comment (GuillaumeGomez) [reply generator.py:38]: wouuuups

Comment (sdroege) [reply generator.py:38]: Not a problem really, but we should remove the Python 2 backwards compat code. This is 2020.

Comment (GuillaumeGomez) [reply generator.py:38]: Absolutely, it's just that I'm so used to write this code that I didn't pay attention. Sending a PR.

**Comment by MarijnS95** [comment]
There's one more thing I have in the pipeline on my gstreamer-rs fork: autoformat this file with [black](https://github.com/psf/black) to make future alterations easier. Anyone interested? Keep in mind this will normalize (currently mixed between `'` and `"`) quotation usage.

**Comment by GuillaumeGomez** [comment]
Please go ahead!

**Comment by MarijnS95** [comment]
@GuillaumeGomez Done! Hope that's the last of 'em, and we can finally get back to writing Rust :)

**Comment by GuillaumeGomez** [comment]
Thanks a lot!

**Pull request merged**